cmd/uniq/uniq.go

```diff
@@ -0,0 +1,148 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+)
+
+type options struct {
+	printDuplicates bool
+	printEmptyLines bool
+	printEveryOnce  bool
+	printLineNumber bool
+}
+
+type Line struct {
+	text *string
+	nums []int
+}
+
+func usage() {
+	fmt.Println(`Usage:
+uniq [[-dup | -every] -empty | -num]`)
+}
+
+func parseArgs(args []string) (options, error) {
+	var opts options
+
+	for _, opt := range args[1:] {
+		switch opt {
+		case "-dup":
+			opts.printDuplicates = true
+		case "-empty":
+			opts.printEmptyLines = true
+		case "-every":
+			opts.printEveryOnce = true
+		case "-num":
+			opts.printLineNumber = true
+		default:
+			usage()
+			return options{}, errors.New("Wrong option")
+		}
+	}
+	if opts.printEveryOnce && opts.printDuplicates {
+		usage()
+		return options{}, errors.New("Choose -dup OR -every")
+	}
+
+	return opts, nil
+}
+
+func addLine(linep *Line, opts options, empAdded bool) (bool, bool) {
+	if *linep.text == "\n" {
+		if opts.printEmptyLines && !empAdded {
+			return true, true
+		}
+	} else if len(linep.nums) == 1 {
+		return true, empAdded
+	}
+
+	return false, empAdded
+}
+
+func scanLines(input io.Reader, opts options) []*Line {
+	reader := bufio.NewReader(input)
+	linesPtrMap := make(map[string]*Line)
```

Comment:
@tiago4orion I believe that we have performance and space improvement using `map[string]*Line` once that [LinesOrdered](https://github.com/c0defellas/enzo/pull/28/files#diff-d2c7a5b53893853781f228b0716c857fR69) which is a []*Line (returned by `scanLines()`, contains only the pointers of the lines scanned into the map instead of pure `Line`. If we attempt to reduce the indirection used in the map, we'll probably have to use more space to build the LinesOrdered slice.

See `perf` results on my machine of uniq as it is now (unfortunately for testing not using pointers I should have to change much of the logic, so I didn't do it).

``` shell
$ cat ~/projs/rsrch/anas.platyrhynchos/data | wc -l
18532826

$ taskset 2 perf stat -e task-clock,cache-references,cache-misses,cycles,instructions,branches,branch-misses \
./uniq_pointer -num -dup < ~/projs/rsrch/anas.platyrhynchos/data
...

 Performance counter stats for './uniq_pointer -num -dup':

      64623,899938      task-clock:u (msec)       #    0,996 CPUs utilized
     6.161.710.231      cache-references:u        #   95,347 M/sec                    (83,33%)
     3.942.617.559      cache-misses:u            #   63,986 % of all cache refs      (83,32%)
   153.774.203.335      cycles:u                  #    2,380 GHz                      (66,68%)
   208.763.595.755      instructions:u            #    1,36  insn per cycle           (83,34%)
    41.554.644.963      branches:u                #  643,023 M/sec                    (83,33%)
       162.322.952      branch-misses:u           #    0,39% of all branches          (83,34%)

      64,867745418 seconds time elapsed

```
These are my concerns about not using pointers:

     6.161.710.231      **cache-references:u**        #   95,347 M/sec                    (83,33%)
     3.942.617.559      **cache-misses:u**            #   63,986 % of all cache refs      (83,32%)

Using pointers we already have a lot of cache-misses. That misses will not increase using larger buffers instead of pointers? We need tests.